### PR TITLE
qutewm

### PR DESCRIPTION
**Pull request by Kingdread**
Much to my delight, it's happening, `qutewm` is coming :tada: :tada: :tada: 

As discussed in #1831, for some tests a small window manager is needed to focus the right window. qutewm should be able to deal with this, then we can use it inside Xvfb to run our tests. Ideally, it should work with the `activateWindow` method from Qt.

qutewm should implement a subset of [ICCCM](https://tronche.com/gui/x/icccm/) and [EWMH](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html) to work with standard tools (and to prevent rolling our own ["talk to the WM"-protocol](https://xkcd.com/927/))

This is not ready yet, but I decided to open it in case anyone is curious or wants to help.

**Note that qutewm is not meant as a standalone window manager. It's a
test tool, nothing more.**<sup><sup><sup>yet</sup></sup></sup>
# Todo
- [x] implement urgency flag handling (should also activate the window)
- [x] test integration with Qt methods ([`activateWindow`](http://doc.qt.io/qt-5/qwidget.html#activateWindow))
- [x] integrate it as a fixture (using `testprocess`, like our webserver)
- [x] logging + display of logging output via pytest (like httpbin)
- [x] tests for qutewm
- [x] @qutewm annotation, replace @no_xvfb with @qutewm
- [x] check for other WMs, skip test if another WM is used
- [x] `:debug-focus-window` command
- [x] remove herbstluftwm from travis
- [x] cleanup, remove the repl
- [x] revert the dependency
# Open questions

> I think I would distinguish between tests that just need “automatically focus new windows” behavioer and tests that specifically need the ability to use “:debug-focus-window”.

If we stick to EWMH, other window managers would also handle our activation request (if we use `_NET_ACTIVE_WINDOW` instead of the urgency flag). Then other window managers could also run the tests, so `@qutewm` would just mean "a window manager should be running, start qutewm if none is available" and we wouldn't need to distinguish between "WM that activates newest" and "qutewm".
# Resources
- [Inter-Client Communication Conventions Manual](https://tronche.com/gui/x/icccm/)
- [Extended Window Manager Hints](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html)
- [python-xlib](https://github.com/python-xlib/python-xlib) (unfortunately, no documentation)
- [python-ewmh](https://github.com/parkouss/pyewmh) (probably more useful as a client, not a WM)
- [xlib manual](https://tronche.com/gui/x/xlib/)

### CODE REVIEW

**Comment by Kingdread** [comment]
Contrary to what the Qt docs say about `activateWindow` 

> This function performs the same operation as clicking the mouse on the title bar of a top-level window. 

this function seems to just input-focus the window. Even in herbstluftwm the "active" border stays around the old window, but the key-presses are redirected to the new window. It seems like Qt just calls `XSetInputFocus` and does not use the EWMH `_NET_ACTIVE_WINDOW` message.

We probably need to use the urgency workaround if we want our manager to handle it.

**Comment by The-Compiler** [comment]
> If we stick to EWMH, other window managers would also handle our activation request

Chances are (especially tiling WMs) aren't going to honour it. I'd rather play safe than sorry, and skip tests needing any WM cooperation on Linux without Xvfb.

**Comment by The-Compiler** [comment]
By the way, [Xephyr](https://www.freedesktop.org/wiki/Software/Xephyr/) might make your life a lot easier while debugging this. I also have a little [script](http://paste.the-compiler.org/view/3994d9bb) to debug herbstluftwm in herbstluftwm - maybe there are some interesting bits in there too.

**Comment by haasn** [comment]
FWIW a default xmonad deliberately ignores `_NET_ACTIVE_WINDOW` because of programs like firefox using it to annoy the user. (You have to both specifically enable ewmh support *and* remove this from the ewmh hint blacklist, for it to work)

So there's one example of a WM for which pretty much nothing you can do will ever determine which window has focus.

**Comment by haasn** [comment]
> Contrary to what the Qt docs say about activateWindow

I'm very sure those docs are just horrendeously inaccurate. It's true on essentially no platform from what I can tell - in most environments (e.g. Windows) it will simply flash the taskbar.

**Comment by Kingdread** [comment]
I agree that we should better be safe by not depending on the host WM to honour our requests, so @qutewm should skip if another WM is running.

Thanks for the script, I've already been using Xephyr, but I'll see if I can adapt it to qutewm and spawn sone things automatically.

I've only tested my local WM, not Xmonad, so I was not aware of those restrictions, but that's good to keep in mind.

I'm not yet sure what the Qt methods do and I'm not very familiar with its source yet, but in the end we're probably going to use the uegency-hint, which *should* be available via Qt.

**Comment by The-Compiler** [comment]
It is - for some reason as [`QApplication::alert`](http://doc.qt.io/qt-5/qapplication.html#alert) and not as a `QWidget` method though.

**Comment by Kingdread** [comment]
Note to myself:

`QApplication::alert` seems to send a [`_NET_WM_STATE`](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472615568) client message with `_NET_DEMANDS_ATTENTION`. It does not use `WM_HINTS`. Kind of confusing, since `_NET_WM_STATE` is EWMH, but `activateWindow` does not use EWMH...

**Comment by Kingdread** [comment]
I'm a bit confused, tox uses `pytest-xvfb` to run everything inside Xvfb. `pytest-xvfb` starts its own Xvfb instance. Yet, in the [Dockerfile](https://github.com/The-Compiler/qutebrowser/blob/master/misc/docker/ubuntu-xenial/Dockerfile#L32) you seem to be starting Xvfb manually, followed by `herbstluftwm`. Doesn't this mean that Xvfb is run twice, once just for `herbstlufwm` and once for the tests? And isn't the `herbstluftwm` useless, as it runs in another X screen? Or is `--no-xvfb` somehow enabled on the docker? Or am I missing something else?

**Comment by The-Compiler** [comment]
Err... look, a bunny! :rabbit: :sweat_smile:  I'm... just as confused as you are.

**Comment by Kingdread** [comment]
Look, a friend for your bunny :rabbit2:

Judging by the [log](https://travis-ci.org/The-Compiler/qutebrowser/jobs/152643377#L3848), `DISPLAY` is actually set to `:1` in the tests (which means it's not using the Xvfb from the Dockerfile, which is `:0`). `herbstclient` times out and doesn't print the layout, which I guess is normal ([herbstclient man page](https://herbstluftwm.org/herbstclient.html)):

> It waits endlessly for a response from herbstluftwm (there is no timeout yet).

(though I don't know if that sentence applies if no herbstluftwm is running on that display).

`qutewm` started successfully in an [earlier run](https://travis-ci.org/The-Compiler/qutebrowser/jobs/152050820#L4487) (aca121b15eef856ee05d4b3807207591d280464c), which means that it started its event loop, which in turn means that it didn't exit early, which in turn means that no other WM was running on that screen.

I'm pretty certain now that both the `Xvfb` and `herbstluftwm` from Docker are not used at all...

**Comment by The-Compiler** [comment]
Makes sense, yeah... In 3ec7a01590ef39ee48f6cb2563aeab741b1401a7 I originally played with using Xdummy as I had some issues with Xvfb, which I then later solved by using docker's `--privileged`. Probably I just translated that Dockerfile from Xdummy to Xvfb at some point and didn't realize what was actually going on inside the container?

**Comment by Kingdread** [comment]
I'll see if I can somehow test this with a modified docker locally, though it'll probably some days (and I need to read a short introduction to docker :wink:). Not that it should matter much (the second Xvfb shouldn't interfere with pytest-xvfb), but this has piqued my interest and it'll probably help avoiding confusion in the future :wink:

And then I can finally remove all those debug commits :laughing:

**Comment by The-Compiler** [comment]
I guess we can simply stop starting Xvfb and herbstluftwm in the Dockerfile and see what happens? Note however that Travis won't use the new Dockerfile automatically, because it gets built via hub.docker.com and then just downloaded on Travis.

**Comment by Kingdread** [comment]
That's another idea, though I didn't want all your things to break if I'm somehow wrong :stuck_out_tongue_winking_eye: But I guess in this case we can just revert the changes.

**Comment by Kingdread** [comment]
#1863 has cleared things up now, so I'm removing the "needs more investigation" tag

File: tests/end2end/fixtures/test_qutewm.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+@pytest.mark.qutewm
+class TestQuteWM:
+
+    # qtbot.wait is used to wait for WAIT_DELAY ms. This makes sure that Qt
+    # processed all events (like the focus changed event from the X server),
+    # so that QMainWindow.isActiveWindow works as expected and returns the
+    # correct result. Waiting for qutewm.window_focused is not enough, as that
+    # may fire when Qt hasn't yet registered the change, so we need to give it
+    # some time and at least enter the event loop once.
+    WAIT_DELAY = 200
```
Comment (Kingdread) [inline tests/end2end/fixtures/test_qutewm.py]: Probably nicer to do this the same way `qutebrowser` does it, i.e. connect to `qapp.focusChanged` and do something like

```python
window = new.window()
if (old is None or old.window() != window) and isinstance(window, OurTestWindow):
    window.window_activated.emit()
```

File: tests/end2end/fixtures/test_qutewm.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        with qtbot.waitSignal(qutewm.window_closed):
+            window.close()
+
+    def test_qutewm_exits_when_wm_running(self, qtbot):
+        new_process = qutewm.QuteWMProcess()
+        new_process.start()
+        new_process.terminate()
+        assert new_process.wm_failed
+
+    def test_focus_on_window_open_and_close(self, qutewm, qtbot):
```
Comment (Kingdread) [inline tests/end2end/fixtures/test_qutewm.py]: Much code duplication, would be nice to refactor this out. The only thing that basically changes is the last action (`.close()` vs `.activateWindow()` vs `qapp.alert()`) and the last focused window...

**Comment by Kingdread** [comment]
I think this is ready for an initial review (once you have some more time). The lint error seems to be unrelated to my changes and fixed in master (553eaee4673b83c5422d22e5511ff5a5ce263a32).

Finishing it correctly is still dependent on The-Compiler/pytest-xvfb#5 (the Xauthority thing, which in turn is probably blocked the `PyVirtualDisplay` fix), as it's currently using my patched version.

**Comment by The-Compiler** [comment]
Let's wait for a few more days I'd say - and if nothing happens, you can always open another PR with [your pytest-xvfb changes](https://github.com/The-Compiler/pytest-xvfb/compare/master...Kingdread:master) and I'll make a new release with that in.

**Comment by The-Compiler** [comment]
It's been three weeks and no response in the PR :cry:

So I'd say let's do the necessary change to `pytest-xvfb` and go ahead with that.

**Comment by Kingdread** [comment]
Yeah, unfortunately it seems like PyVirtualDisplay is not as actively maintained as `qutebrowser` or `pytest-xvfb`, so I will prepare a PR for `pytest-xvfb` with the required changes, and then I'll rebase and force-push this one.

**Comment by Kingdread** [comment]
Just did a huge rebase onto master to resolve the conflicts, while also taking out the dependency on my `pytest-xvfb` fork (thanks to ac4724807a12c82e74ce9b0d7c4592921bac6044). I hope nothing broke :wink:

**Comment by The-Compiler** [comment]
Observations so far:

- Web inspector tests don't work with QtWebEngine. I guess ":fake-key sending key to the website with other window focused" could use `:debug-console` instead, and the inspector tests just need a `@qtwebengine_skip` as we can't easily test stuff with the restart.
- On OS X, "Using new-instance-open-target.window = first-opened" from `invoke.feature` fails:

```
INVALID: Traceback (most recent call last):
INVALID:   File "/Users/travis/build/The-Compiler/qutebrowser/qutebrowser/app.py", line 104, in <lambda>
INVALID:     target_arg=target_arg))
INVALID:   File "/Users/travis/build/The-Compiler/qutebrowser/qutebrowser/app.py", line 250, in process_pos_args
INVALID:     win_id = mainwindow.get_window(via_ipc, force_tab=True)
INVALID:   File "/Users/travis/build/The-Compiler/qutebrowser/qutebrowser/mainwindow/mainwindow.py", line 88, in get_window
INVALID:     window.setWindowState(window.windowState() & ~Qt.WindowMinimized)
INVALID: RuntimeError: wrapped C/C++ object of type MainWindow has been deleted
```

Not sure yet what that is...

**Comment by Kingdread** [comment]
You mean we can't do tests with the webinspector and QtWebEngine, as we'd have to basically enable developer extras, restart qutebrowser (and keep the basedir) and then continue the test? Otherwise the inspector won't work?

For Mac OS, I need to do some more investigating, as `qutewm` shouldn't even run there, so it shouldn't have any impact...

**Comment by The-Compiler** [comment]
> You mean we can't do tests with the webinspector and QtWebEngine, as we'd have to basically enable developer extras, restart qutebrowser (and keep the basedir) and then continue the test? Otherwise the inspector won't work?

Right. Or we need to find some other way to enable the webinspector earlier (so qutebrowser can set the environment variable before reading the config).

**Comment by Kingdread** [comment]
Strange, the Mac test seems to pass now even though I didn't touch the Mac code in the meantime... 🤔

**Comment by The-Compiler** [comment]
Fun... it sounds like another race condition caused by the current `objreg` implementation which tries to automatically clean windows up once they're destroyed, which is... troublesome.

**Comment by The-Compiler** [comment]
I kinda lost track here - what's the state of this, @Kingdread?

**Comment by Kingdread** [comment]
I think the "needs more investigation" was because of the `pytest-xvfb`/`PyVirtualDisplay` changes (which is now done), so I guess this is ready for an initial review (thought I probably need to update the requirements to reflect that). The tests seem to work fine.

**Comment by The-Compiler** [comment]
You'll get #2054 to review in return :laughing:

Might be next week until I have time to review this though.

File: tests/end2end/fixtures/qutewm_sub.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    def _wm_running(self, error, request):
+        """Called when another WM is already running."""
+        log.error("Another window manager is running, exiting")
+        # This is called async, which means we can't just raise an exception,
+        # we need to signal the main thread to stop.
+        self._retcode = 42
+
+    def _error_handler(self, error, request):
+        # This is called async, which means we can't just raise an exception,
+        # we need to signal the main thread to stop.
```
Comment (The-Compiler) [inline tests/end2end/fixtures/qutewm_sub.py]: So do we do that somewhere

File: tests/end2end/fixtures/qutewm_sub.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            self.atoms.supported,
+            self.atoms.active_window,
+            self.atoms.client_list,
+            self.atoms.wm_state,
+        ]
+        self.root.change_property(
+            self.atoms.supported,
+            Xatom.ATOM,
+            32,
+            attributes,
```
Comment (The-Compiler) [inline tests/end2end/fixtures/qutewm_sub.py]: nitpick: I'd prefer the closing `)` here and no dangling `,` for function calls (same below, but won't comment that everywhere)

File: tests/end2end/fixtures/qutewm_sub.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        )
+
+    def _set_supporting_wm_check(self):
+        """Create and set a window for _NET_SUPPORTING_WM_CHECK."""
```
Comment (The-Compiler) [inline tests/end2end/fixtures/qutewm_sub.py]: Can you explain what this is for here?

File: tests/end2end/fixtures/qutewm_sub.py [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+        """Start the X event loop.
+
+        Return:
+            The manager's exit code.
+        """
+        if self._retcode is not None:
+            # avoid the "event loop started" message if we exit anyway
+            return self._retcode
+        log.info("event loop started")
+        while 1:
```
Comment (The-Compiler) [inline tests/end2end/fixtures/qutewm_sub.py]: `while True:`

File: tests/end2end/fixtures/qutewm_sub.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+            if self._retcode is not None:
+                return self._retcode
+            try:
+                ev = self.root.display.next_event()
+                handler = self._handlers.get(ev.type)
+                if handler:
+                    handler(ev)
```
Comment (The-Compiler) [inline tests/end2end/fixtures/qutewm_sub.py]: Would it be useful to log unhandled events here?

File: tests/end2end/fixtures/qutewm_sub.py [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+                            value_mask=ev.value_mask)
+
+    def _on_circulate_request(self, ev):
+        """Called when a CirculateRequest is intercepted."""
+        ev.window.circulate(ev.place)
+
+    def _on_unmap_notify(self, ev):
+        """Called when a window is unmapped from the screen."""
+        log.debug("window unmapped: {}".format(ev.window))
+        if ev.event == self.root and not ev.from_configure:
```
Comment (The-Compiler) [inline tests/end2end/fixtures/qutewm_sub.py]: Should this be `==` or `is`?

File: tests/end2end/fixtures/qutewm_sub.py [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+        # re-order window_stack so that the active window is at
+        # window_stack[-1]
+        try:
+            index = self.window_stack.index(window)
+        except ValueError:
+            # Okay, fine then
+            pass
+        else:
+            self.window_stack = (self.window_stack[index + 1:] +
+                                 self.window_stack[:index + 1])
```
Comment (The-Compiler) [inline tests/end2end/fixtures/qutewm_sub.py]: Wouldn't it be easier to pop/re-append the window from/to the stack?

File: tests/end2end/fixtures/qutewm_sub.py [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+                                                   Xatom.ATOM, 0, 32)
+        if client_properties is None:
+            client_properties = set()
+        else:
+            client_properties = set(client_properties.value)
+
+        action = ev.data[1][0]
+        updates = {ev.data[1][1]}
+        if ev.data[1][2] != 0:
+            updates.add(ev.data[1][2])
```
Comment (The-Compiler) [inline tests/end2end/fixtures/qutewm_sub.py]: What's this data?

File: tests/end2end/fixtures/test_qutewm.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+
+@pytest.mark.qutewm
+class TestQuteWM:
+
+    @classmethod
+    def get_window(cls):
+        """Return a window suitable for testing."""
+        window = Window()
+        return window
```
Comment (The-Compiler) [inline tests/end2end/fixtures/test_qutewm.py]: What's the point of this classmethod? Why not just use `Window()` instead of `self.get_window()`?

File: tests/end2end/fixtures/test_qutewm.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        window = self.get_window()
+        qtbot.addWidget(window)
+        with qtbot.waitSignals([qutewm.window_opened, qutewm.window_focused,
+                                window.activated]):
+            window.show()
+        assert window.isActiveWindow()
+        with qtbot.waitSignal(qutewm.window_closed):
+            window.close()
+
+    def test_qutewm_exits_when_wm_running(self, qtbot):
```
Comment (The-Compiler) [inline tests/end2end/fixtures/test_qutewm.py]: Will this test fail with no WM running?

File: tests/end2end/fixtures/qutewm_sub.py [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+
+Exit codes:
+    42 - another window manager is running
+"""
+
+import sys
+import signal
+import logging
+
+from Xlib.display import Display
```
Comment (The-Compiler) [inline tests/end2end/fixtures/qutewm_sub.py]: I'd prefer importing the `display` module and using `display.Display`

File: qutebrowser/mainwindow/mainwindow.py [hunk 14]
```diff
@@ -472,6 +472,17 @@ def showEvent(self, e):
...
+        """Extend changeEvent to register an activationChange.
+
+        Args:
+            e: The QChangeEvent
+        """
+        super().changeEvent(e)
+        if e.type() == QEvent.ActivationChange and self.isActiveWindow():
+            log.misc.debug("Window activation changed to {!r} (win_id={})"
+                           .format(self, self.win_id))
+
```
Comment (Kingdread) [inline qutebrowser/mainwindow/mainwindow.py]: I think I tried that at first, but it didn't seem to work reliably. Either there is a difference in the order in which those events are processed (so we signal that activation has changed, but `window.isActivated` still returns false), or some other things aren't right. I figured just using the right event would probably be easier.

File: tests/end2end/fixtures/qutewm.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+            self.ready.emit()
+        else:
+            match = self._event_re.search(line)
+            if match:
+                signal = {
+                    'created': self.window_opened,
+                    'closed': self.window_closed,
+                    'activated': self.window_focused,
+                }[match.group(1)]
+                signal.emit(int(match.group(2), 16))
```
Comment (Kingdread) [inline tests/end2end/fixtures/qutewm.py]: You mean something like `'signal: activated'`  is parsed and when `'signal: foobar'` is encountered we raise an exception ("Invalid signal from qutewm")?

File: tests/end2end/fixtures/qutewm_sub.py (hunk 3 above)
Comment (Kingdread) [inline tests/end2end/fixtures/qutewm_sub.py]: I actually think this isn't quite right anymore, I've switched to the `onerror` parameter in `__init__` and just let qutewm continue to run, even on errors. I'll have to change that again.

File: tests/end2end/fixtures/qutewm_sub.py (hunk 5 above)
Comment (Kingdread) [inline tests/end2end/fixtures/qutewm_sub.py]: It's just a window that is specified by EWMH, to make sure the window manager supports EWMH. I can add a comment with the explanation, or at least a link to the [spec](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472693600).

File: tests/end2end/fixtures/qutewm_sub.py (hunk 7 above)
Comment (Kingdread) [inline tests/end2end/fixtures/qutewm_sub.py]: Probably not, X can generate lots of events, it would probably just clutter the log. At least I'd hide it behing some verbose flag, which you can turn on when you really want it (probably only when you're working on qutewm itself).

File: tests/end2end/fixtures/qutewm_sub.py (hunk 9 above)
Comment (Kingdread) [inline tests/end2end/fixtures/qutewm_sub.py]: That would indeed be easier :laughing:

File: tests/end2end/fixtures/qutewm_sub.py (hunk 10 above)
Comment (Kingdread) [inline tests/end2end/fixtures/qutewm_sub.py]: The zeroeth index is the action (add, remove, toggle), the first index is the property, and the second may contain another property or just be empty. Again, I will add a comment and a link to the relevant [spec section](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472615568):

```
  window  = the respective client window
  message_type = _NET_WM_STATE
  format = 32
  data.l[0] = the action, as listed below
  data.l[1] = first property to alter
  data.l[2] = second property to alter
  data.l[3] = source indication
  other data.l[] elements = 0
```

File: tests/end2end/fixtures/test_qutewm.py (hunk 11 above)
Comment (Kingdread) [inline tests/end2end/fixtures/test_qutewm.py]: Old remnants when I had more set-up to a window when I tried to get the "Activation changed" signal.

File: tests/end2end/fixtures/test_qutewm.py (hunk 12 above)
Comment (Kingdread) [inline tests/end2end/fixtures/test_qutewm.py]: Well, the whole class is marked with "qutewm", so there should be qutewm running (or the thing is skipped). Then the test tries to start a second instance, which should fail.

File: tests/end2end/fixtures/qutewm_sub.py (hunk 7 above)
Comment (The-Compiler) [inline tests/end2end/fixtures/qutewm_sub.py]: Maybe just re-add the debug logging a few lines below?

File: tests/end2end/fixtures/qutewm.py (hunk 15 above)
Comment (Kingdread) [inline tests/end2end/fixtures/qutewm.py]: X errors should make qutewm exit anyway, which is checked (the "ProcessExited" error). Not sure about warnings (does X even have warnings which are not errors?), and if they should be fatal.

**Comment by Kingdread** [comment]
Not sure where those `QXcbConnection` errors come from, they seem to arise randomly locally. Need some more investigation.

**Comment by The-Compiler** [comment]
@Kingdread did you ever figure out what's wrong here?

**Comment by Kingdread** [comment]
Unfortunately, no. The errors seemed random and didn't appear on every system, and I wasn't really able to debug them. It's possible that they've been fixed by now, but it's also possible that it's just an error in the implementation of qutewm and still persists.

Though I can try and bring this up-to-date and check again what happens.

**Comment by The-Compiler** [comment]
Would be awesome if you could at least rebase it on the latest master so we can see what happens there.

**Comment by Kingdread** [comment]
Rebased and pushed, but it seems like the error still persists. Also, the `test_qutewm`-test seems to be flaky... Let's see what Travis says!

**Comment by The-Compiler** [comment]
I'm assuming this is abandoned, and unfortunately I don't have the time to dig into these errors at the moment. I'm closing it, but I added the patch as a comment to #1831 to ensure nothing gets lost.

Thanks for the contribution anyways!
